src/Servant/API.hs

```diff
@@ -69,3 +77,29 @@ import           Servant.API.ReqBody      (ReqBody)
 import           Servant.API.Sub          ((:>))
 import           Servant.Utils.Links      (HasLink (..), IsElem, IsElem',
                                            URI (..), safeLink)
+
+-- | Turn an API type into its canonical form.
+--
+-- The canonical form is defined and will basically turn:
+--
+-- > "hello" :> (Get Hello :<|> ReqBody Hello :> Put Hello)
+--
+-- into
+--
+-- > ("hello" :> Get Hello) :<|> ("hello" :> ReqBody Hello :> Put Hello)
+--
+-- i.e distributing all ':>'-separated bits into the subsequent ':<|>'s.
+type family Canonicalize api :: * where
+  -- requires UndecidableInstances
+  Canonicalize (a :> (b :<|> c)) = a :> Canonicalize b :<|> a :> Canonicalize c
+  Canonicalize ((a :<|> b) :> c) = a :> Canonicalize c :<|> b :> Canonicalize c
+  Canonicalize (a :> b)          = Redex b (Canonicalize b) a
+  Canonicalize (a :<|> b)        = Canonicalize a :<|> Canonicalize b
+  Canonicalize a                 = a
+
+type family Redex a b c :: * where
+  Redex a a first = Canonicalize first :> a
+  Redex a b first = Canonicalize (first :> b)
+
+canonicalize :: Canonicalize layout ~ t => Proxy layout -> Proxy t
```

Comment:
Oh of course, forgot to put that back to `Proxy layout -> Proxy (Canonicalize layout)`. This was just the result of an intermediary implementation of the canonicalization process. I'm about to push this along with some more docs and a changelog entry.